File: Algorithms/0097_Interleaving_String/Java/InterleavingStringSolution2.java

// Space: O(n)
// Time: O(n)

class Solution {
    public boolean isInterleave(String s1, String s2, String s3) {
        if (s1.length() == s2.length() && s2.length() == s3.length() && s3.length() == 0) return true;
        if (s1.length() + s2.length() != s3.length()) return false;
        s1 = "#" + s1;
        s2 = "#" + s2;
        s3 = "#" + s3;
        boolean[][] dp = new boolean[s1.length()][s2.length()];
        for (boolean[] row : dp) {
            Arrays.fill(row, false);
        }

        for (int y = 0; y < s1.length(); y++) {
            for (int x = 0; x < s2.length(); x++) {
                if (x == 0 && y == 0) dp[y][x] = true;
                else if (y == 0 && s2.charAt(x) == s3.charAt(x + y) && dp[y][x - 1]) dp[y][x] = true;
                else if (x == 0 && s1.charAt(y) == s3.charAt(x + y) && dp[y - 1][x]) dp[y][x] = true;
                else if (s2.charAt(x) == s3.charAt(x + y) && dp[y][x - 1]) dp[y][x] = true;
                else if (s1.charAt(y) == s3.charAt(x + y) && dp[y - 1][x]) dp[y][x] = true;
            }
        }
        return dp[s1.length() - 1][s2.length() - 1];
    }
}

File: Algorithms/0097_Interleaving_String/Python/Interleaving_String_Solution_2.py

# Space: O(n)
# Time: O(n)

class Solution:
    def isInterleave(self, s1, s2, s3):
        length_1, length_2, length_3 = len(s1), len(s2), len(s3)
        if length_1 == length_2 == length_3 == 0: return True
        if length_1 + length_2 != length_3: return False

        s1 = '#' + s1
        s2 = '#' + s2
        s3 = '#' + s3

        dp = [[False for _ in range(length_1 + 1)] for _ in range(length_2 + 1)]

        for y in range(len(dp)):
            for x in range(len(dp[0])):
                if x == 0 and y == 0:
                    dp[y][x] = True
                elif y == 0 and s1[x] == s3[y + x] and dp[y][x - 1] is True:
                    dp[y][x] = True
                elif x == 0 and s2[y] == s3[y + x] and dp[y - 1][x] is True:
                    dp[y][x] = True
                elif s1[x] == s3[y + x] and dp[y][x - 1] is True:
                    dp[y][x] = True
                elif s2[y] == s3[y + x] and dp[y - 1][x] is True:
                    dp[y][x] = True
        return dp[-1][-1]




